pkg/spanfilter/policymatch/policymatch.go

```diff
@@ -0,0 +1,282 @@
+	v1_common "github.com/grafana/tempo/pkg/tempopb/common/v1"
+	v1_trace "github.com/grafana/tempo/pkg/tempopb/trace/v1"
+	"github.com/grafana/tempo/pkg/traceql"
+)
+
+// PolicyMatch is a set of attributes that must match a span for the span to match the policy.
+type PolicyMatch struct {
+	attributes []MatchPolicyAttribute
+}
+
+// NewPolicyMatch returns a new PolicyMatch with the given attributes. If no attributes are given, then the policy matches all spans.
+func NewPolicyMatch(attrs ...MatchPolicyAttribute) *PolicyMatch {
+	return &PolicyMatch{attributes: attrs}
+}
+
+func (p *PolicyMatch) AddAttr(attr MatchPolicyAttribute) {
+	p.attributes = append(p.attributes, attr)
+}
+
+// MatchAttrs returns true if all attributes in the policy match the attributes in the span.  String, bool, int, and floats are supported.  Regex MatchType may be applied to string span attributes.
+func (p *PolicyMatch) MatchAttrs(attrs []*v1_common.KeyValue) bool {
+	// If the policy has no attributes, then it matches.
+	if len(p.attributes) == 0 {
+		return true
+	}
+
+	// If the span has no attributes, then it does not match.
+	if len(attrs) == 0 {
+		return false
+	}
+
+	matches := 0
+	for _, pa := range p.attributes {
+		key := pa.Key()
+		for _, attr := range attrs {
+			if attr.GetKey() == key {
+				v := attr.Value.Value
+
+				// For each type of value, check if the policy attribute value matches the span attribute value.
+				switch value := v.(type) {
+				case *v1_common.AnyValue_StringValue:
+					if !pa.MatchString(value.StringValue) {
+						return false
+					}
+					matches++
+				case *v1_common.AnyValue_IntValue:
+					if !pa.MatchInt64(value.IntValue) {
+						return false
+					}
+					matches++
+				case *v1_common.AnyValue_DoubleValue:
+					if !pa.MatchFloat64(value.DoubleValue) {
+						return false
+					}
+					matches++
+				case *v1_common.AnyValue_BoolValue:
+					if !pa.MatchBool(value.BoolValue) {
+						return false
+					}
+					matches++
+				}
+			}
+		}
+	}
+
+	return len(p.attributes) == matches
+}
+
+// MatchIntrinsicAttrs returns true when all intrinsic values in the policy match the span.
+func (p *PolicyMatch) MatchIntrinsicAttrs(span *v1_trace.Span) bool {
+	matches := 0
+
+	for _, pa := range p.attributes {
+		switch pa.KeyIntrinsic() {
+		// case traceql.IntrinsicDuration:
+		// case traceql.IntrinsicChildCount:
+		// case traceql.IntrinsicParent:
+		case traceql.IntrinsicName:
+			if !pa.MatchString(span.GetName()) {
+				return false
+			}
+			matches++
+		case traceql.IntrinsicStatus:
+			if !pa.MatchStatusCode(span.GetStatus().GetCode()) {
+				return false
+			}
+			matches++
+		case traceql.IntrinsicKind:
+			if !pa.MatchSpanKind(span.GetKind()) {
+				return false
+			}
+			matches++
+		}
+	}
+
+	return len(p.attributes) == matches
+}
+
+// MatchPolicyAttribute is an attribute that must match a span for the span to match the policy.
+type MatchPolicyAttribute interface {
+	Key() string
```

Comment:
i'm trying to think of a way to simplify this interface. i think it will benefit the code using the MatchPolicyAttributes.

Could it be as simple as:
```
type matchPolicyAttribute interface {
  matchesSpan(*span)    // used by MatchIntrinsicAttrs
  matchesAttribute(*attr)  // used by MatchAttrs
}
```

wdyt?

Reply:
Very good suggestion. I will do an update to this PR.

Reply:
I took  a bit different direction, but with the same idea in mind. I have created a separate structure IntrinsicPolicyMatch to perform matches span's intrinsic attributes.

Reply:
Attaching benchmarks here.
```
goos: darwin
goarch: arm64
pkg: github.com/grafana/tempo/pkg/spanfilter
                                        │    o.txt     │                n.txt                 │
                                        │    sec/op    │    sec/op     vs base                │
SpanFilter_applyFilterPolicyNone-8        4.447m ±  9%   4.444m ±  2%        ~ (p=0.436 n=10)
SpanFilter_applyFilterPolicySmall-8       57.37m ± 10%   54.05m ±  4%   -5.79% (p=0.000 n=10)
SpanFilter_applyFilterPolicyMedium-8      92.36m ±  7%   54.05m ± 14%  -41.48% (p=0.000 n=10)
SpanFilter_applyFilterPolicyRegex-8       67.96m ±  4%   53.47m ±  7%  -21.33% (p=0.000 n=10)
SpanFilter_applyFilterPolicyIntrinsic-8   28.64m ±  2%   22.68m ±  2%  -20.84% (p=0.000 n=10)
geomean                                   34.06m         27.50m        -19.26%

```

Reply:
nice!